tests/test-dt.cpp

```diff
@@ -41,6 +41,43 @@ class DTTests : public ::testing::Test,
   Sort intsort;
 };
 
+TEST_P(DTTests, DeclareSimpleList)
+{
+  SolverConfiguration sc = GetParam();
+  if (sc.is_logging_solver || s->get_solver_enum() == GENERIC_SOLVER)
+  {
+    return;
+  }
+
+  DatatypeDecl listSpec = s->make_datatype_decl("list");
+  DatatypeConstructorDecl nildecl = s->make_datatype_constructor_decl("nil");
+  DatatypeConstructorDecl consdecl = s->make_datatype_constructor_decl("cons");
+  s->add_selector(consdecl, "head", s->make_sort(INT));
+  s->add_selector_self(consdecl, "tail");
+  s->add_constructor(listSpec, nildecl);
+  s->add_constructor(listSpec, consdecl);
+  Sort listsort = s->make_sort(listSpec);
+}
+
+TEST_P(DTTests, DeclareSimpleListWithForwardRef)
+{
+  SolverConfiguration sc = GetParam();
+  if (sc.is_logging_solver || s->get_solver_enum() == GENERIC_SOLVER)
+  {
+    return;
+  }
+
+  DatatypeDecl listSpec = s->make_datatype_decl("list");
+  Sort forward_ref_listsort = s->make_sort("list", 0);
```

Comment:
Maybe not necessary for this PR but what do you think about adding a `make_forward_ref` convenience function that takes a `DatatypeDecl` as an argument and does essentially this? The advantage would be that the user can't get the names wrong.

Reply:
That's a really good idea! It seems more "future-proof" as well since not every solver necessarily uses uninterpreted sorts for this purpose.